src/EditorFeatures/Core/GoToNextAndPreviousMethodCommandHandler.cs

```diff
@@ -0,0 +1,180 @@
+
+        public void ExecuteCommand(GoToNextMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: true);
+
+        public CommandState GetCommandState(GoToPreviousMethodCommandArgs args, Func<CommandState> nextHandler)
+            => GetCommandState(args.SubjectBuffer, nextHandler);
+
+        public void ExecuteCommand(GoToPreviousMethodCommandArgs args, Action nextHandler)
+            => ExecuteCommand(nextHandler, args.SubjectBuffer, args.TextView, next: false);
+
+        private static CommandState GetCommandState(ITextBuffer subjectBuffer, Func<CommandState> nextHandler)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var documentSupportsSuggestionService = document.Project.Solution.Workspace.Services.GetService<IDocumentSupportsSuggestionService>();
+            if (document == null || !document.SupportsSyntaxTree || !documentSupportsSuggestionService.SupportsRefactorings(document))
+            {
+                return nextHandler();
+            }
+
+            return CommandState.Available;
+        }
+
+
+        private void ExecuteCommand(Action nextHandler, ITextBuffer subjectBuffer, ITextView textView, bool next)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null || !document.SupportsSyntaxTree)
+            {
+                nextHandler();
+                return;
+            }
+
+            var caretPoint = textView.GetCaretPoint(subjectBuffer);
+            if (!caretPoint.HasValue)
+            {
+                nextHandler();
+                return;
+            }
+
+            int? targetPosition = null;
+            var waitResult = _waitIndicator.Wait("", allowCancel: true, action: waitContext =>
+            {
+                targetPosition = GetTargetPosition(document, caretPoint.Value.Position, next, waitContext.CancellationToken);
+            });
+
+            if (waitResult == WaitIndicatorResult.Canceled || targetPosition == null)
+            {
+                return;
+            }
+
+            textView.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(subjectBuffer.CurrentSnapshot, targetPosition.Value), _outliningManagerService);
+        }
+
+        /// <summary>
+        /// Internal for testing purposes.
+        /// </summary>
+        internal static int? GetTargetPosition(Document document, int caretPosition, bool next, CancellationToken cancellationToken)
+        {
+            var syntaxFactsService = document.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFactsService == null)
+            {
+                return null;
+            }
+
+            var root = document.GetSyntaxRootAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+            var members = syntaxFactsService.GetMethodLevelMembers(root);
```

Comment:
There is a test to cover this, but VS2013 navigates to the first member of the next type, and I think we should do the same.